porcupine/plugins/anchors.py

```diff
@@ -0,0 +1,173 @@
+from porcupine.plugins.linenumbers import LineNumbers
+
+# Dependent on code from linenumbers.py
+setup_after = ["linenumbers"]
+
+
+class AnchorManager:
+    def __init__(self, tab_textwidget: tkinter.Text, linenumbers: tkinter.Canvas) -> None:
+        self.tab_textwidget = tab_textwidget
+        self.linenumbers = linenumbers
+
+        self.anchor_symbol = "¶"
+
+        # tkinter have default marks such as "insert", "current", "tkinter::anchor1"
+        self.custom_anchor_prefix = "anchor_"
+
+        linenumbers.bind("<<Updated>>", self.update_linenumbers, add=True)
+
+    def _get_anchors(self) -> List:
+        return [
+            anchor
+            for anchor in self.tab_textwidget.mark_names()
+            if anchor.startswith(self.custom_anchor_prefix)
+        ]
+
+    def _get_cursor_index(self):
+        return self.tab_textwidget.index("insert linestart")
+
+    def toggle_on_off(self, event: tkinter.Event[tkinter.Misc]) -> None:
+
+        self.prevent_duplicate_anchors()
+
+        cursor_index = self._get_cursor_index()
+        for anchor in self.tab_textwidget.mark_names():
+            if (
+                anchor.startswith(self.custom_anchor_prefix)
+                and self.tab_textwidget.index(anchor) == cursor_index
+            ):
+                self.tab_textwidget.mark_unset(anchor)
+                self.linenumbers.do_update()
+                return
+        self.tab_textwidget.mark_set(
+            self.custom_anchor_prefix + str(time.time()), "insert linestart"
+        )
+
+        self.linenumbers.do_update()
+
+    def jump_to_next(self, event: tkinter.Event[tkinter.Misc]) -> None:
+        cursor_row = self._get_cursor_index().split(".")[0]
+        anchor_list = self._get_anchors()
+        anchor_rows = reversed(
+            sorted(
+                [
+                    int(self.tab_textwidget.index(anchorpoint).split(".")[0])
+                    for anchorpoint in anchor_list
+                ]
+            )
+        )
+
+        next_anchor_row = 0
+        for anchor_row in anchor_rows:
+            if anchor_row > int(cursor_row):
+                next_anchor_row = anchor_row
+            else:
+                break
+
+        if not next_anchor_row:
+            return
+        else:
+            self.tab_textwidget.mark_set("insert", f"{str(next_anchor_row)}.0")
+
+        # If cursor is below last row
+        if int(self._get_cursor_index().split(".")[0]) > int(
+            self.tab_textwidget.index(f"@0,{self.tab_textwidget.winfo_height()}").split(".")[0]
+        ):
+            self.tab_textwidget.see("insert")
+
+        # text.see("insert")
+
+        # first_line = int(self._textwidget.index("@0,0").split(".")[0])
+        # last_line = int(
+        #     self._textwidget.index(f"@0,{self._textwidget.winfo_height()}").split(".")[0]
+        # )
+
+        # TODO: If user jumps to the last anchor and then to next, it should jump to the first one in the file.
```

Comment:
This is obviously a matter of opinion, but I would prefer it to do nothing. If I press the key binding to jump forward, I really wouldn't expect to ever jump backwards to the beginning.

Reply:
If you ctrl + f, and you "find next" all the way to the end of the file, and then "find next" again.. doesn't it jump back up?

Reply:
It does, but to me it's fundamentally different from anchors. With Ctrl+F, you typically want to walk through the whole file, making sure you didn't miss anything. With anchors, you just want to jump to a different place.

But it makes sense if you are using anchors as basically a TODO list, for example: you mark many places, and then go through every anchor until there's no more anchors left. Is this something you want to do? Maybe we could add a checkbox to the settings dialog, and then argue about whether it should be checked by default?

Reply:
Haha, that sounds good.

Reply:
https://akuli.github.io/porcupine/settings.html#the-settings-dialog

There's `add_entry`, `add_combobox`, `add_spinbox` and `add_label`, but no `add_checkbutton`. I'll create an `add_checkbutton` method for you.